File: tests/workflow/producer/rust/src/producer/beacons/beacons_shutdownserver.rs
use super::{identification, producer::Control, protocol, scope::Scope, Context};
use crate::stop;
use clibri::server;
use std::pin::Pin;
use tokio::{
    task,
    time::{sleep, Duration},
};

#[allow(unused_variables)]
pub async fn emit<E: server::Error, C: server::Control<E>>(
    beacon: &protocol::Beacons::ShutdownServer,
    scope: &mut Scope<'_, E, C>,
) -> Result<(), String> {
    scope.context.stats.remove(&scope.identification.uuid());
    let control = scope.control.clone();
    scope.deferred(Box::pin(async move {
        if let Err(err) = control.shutdown(false).await {
            stop!("Fail to shutdown; error: {:?}", err);
        }
    }));
    Ok(())
}

File: tests/workflow/producer/typescript/src/producer/beacons/beacons.shutdownserver.ts
import {
	Context,
	Producer,
	Identification,
	Filter,
	Protocol,
} from "../implementation/beacons";
import { panic } from "../../tools";
import { Scope } from "../implementation/scope";

export function emit(
	beacon: Protocol.Beacons.ShutdownServer,
	scope: Scope
): Promise<void> {
	scope.context.ignore(scope.consumer.uuid());
	scope.deferred(() => {
		setTimeout(() => {
			scope.producer.destroy().catch((error: Error) => {
				panic(`Fail to destroy producer: ${error.message}`);
			});
		}, 1000);
	});
	return Promise.resolve();
}
